app/src/main/java/com/github/se/icebreakrr/utils/PermissionManager.kt

```diff
@@ -0,0 +1,116 @@
+package com.github.se.icebreakrr.utils
+
+import android.Manifest
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.content.pm.PackageManager
+import android.util.Log
+import androidx.activity.ComponentActivity
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.core.content.ContextCompat
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+
+class PermissionManager(private val context: Context) {
+
+  lateinit var permissionLauncher: ActivityResultLauncher<Array<String>>
+  private val _permissionStatuses =
+      MutableStateFlow<Map<String, PermissionStatus>>(initialPermissionStatuses())
+  val permissionStatuses: StateFlow<Map<String, PermissionStatus>> = _permissionStatuses
+
+  private val scope = CoroutineScope(Dispatchers.Main)
+
+  // BroadcastReceiver to detect permission changes
+  private val permissionReceiver =
+      object : BroadcastReceiver() {
+        override fun onReceive(context: Context, intent: Intent) {
+          updateAllPermissions()
+        }
+      }
+
+  /**
+   * Initializes the `PermissionManager` by registering a `BroadcastReceiver` to listen for
+   * configuration changes and periodically checks permission statuses every 5 seconds to keep them
+   * up to date.
+   */
+  init {
+    // Register the BroadcastReceiver
+    context.registerReceiver(permissionReceiver, IntentFilter(Intent.ACTION_CONFIGURATION_CHANGED))
```

Comment:
You're absolutely right; there's no point in keeping the periodic check every 5 seconds. I initially implemented this solution before adding the BroadcastReceiver and didn’t remove it since it was useful for debugging the app. But it no longer serves any purpose, so I’ll remove it. This will also resolve the coroutine testing issues.

Reply:
After investigation, I discovered that my BroadcastReceiver wasn’t being triggered during permission changes, as Intent.ACTION_CONFIGURATION_CHANGED doesn’t get triggered by such changes. I couldn’t find a way to detect a permission change directly. To avoid implementing a costly check every 5 seconds, I chose this solution:

There are only two scenarios for permission changes. In-app, via a popup, where the permissionLauncher updates the permission list; and outside the app, via the settings. For the latter, I added a call to updateAllPermissions() in the onResume() function of the MainActivity, ensuring that permissions are up-to-date when returning to the app. This approach covers all cases without needing to use a coroutine and ensures that the permission list is always up-to-date.

Reply:
Great nice solution thank you!